lightning/src/ln/channelmanager.rs

```diff
@@ -3244,6 +3283,27 @@ impl<ChanSigner: ChannelKeys, M: Deref, T: Deref, K: Deref, F: Deref, L: Deref>
 		self.latest_block_height.fetch_sub(1, Ordering::AcqRel);
 		*self.last_block_hash.try_lock().expect("block_(dis)connected must not be called in parallel") = header.block_hash();
 	}
+
+	/// Blocks until ChannelManager needs to be persisted or a timeout is reached. It returns a bool
+	/// indicating whether persistence is necessary.
+	/// Note that the feature `allow_wallclock_use` must be enabled to use this function.
+	#[cfg(any(test, feature = "allow_wallclock_use"))]
+	pub fn wait_timeout(&self, max_wait: Duration) -> bool {
+		self.persistence_notifier.wait_timeout(max_wait)
+	}
+
+	/// Blocks until ChannelManager needs to be persisted.
+	pub fn wait(&self) {
```

Comment:
May be worth noting here and above that only one waiter is woken at a time for each event (because of the re-setting the guard bool to false in wait in the PersistenceNotifier).

Reply:
Hm I tested with multiple listeners just now and all were notified, but maybe it's unreliable and it's just not showing up in testing?

Reply:
Right, on a multi-core host you may get lucky and the threads both step basically at the same rate, but it's not guaranteed.

Reply:
Wait, no, huh? It swaps the wakeup bool back to false before returning. Can you share your test?

Reply:
I added this to `test_background_processor`:
```
diff --git a/background-processor/src/lib.rs b/background-processor/src/lib.rs
index 5d975eb9..7bc76343 100644
--- a/background-processor/src/lib.rs
+++ b/background-processor/src/lib.rs
@@ -203,8 +203,11 @@ mod tests {

                // Initiate the background processors to watch each node.
                let data_dir = nodes[0].persister.get_data_dir();
+               let data_dir1 = nodes[1].persister.get_data_dir();
                let callback = move |node| FilesystemPersister::persist_manager(data_dir.clone(), node);
+               let callback1 = move |node| FilesystemPersister::persist_manager(data_dir1.clone(), node);
                BackgroundProcessor::start(callback, nodes[0].node.clone(), nodes[0].logger.clone());
+               BackgroundProcessor::start(callback1, nodes[0].node.clone(), nodes[0].logger.clone());

                // Go through the channel creation process until each node should have something persisted.
                let tx = open_channel!(nodes[0], nodes[1], 100000);
@@ -239,6 +242,10 @@ mod tests {
                        if !nodes[0].node.get_persistence_condvar_value() { break }
                }

+               let filepath1 = get_full_filepath("test_background_processor_persister_1".to_string(), "manager".to_string());
+               let mut expected_bytes = Vec::new();
+               check_persisted_data!(nodes[0].node, filepath1.clone(), expected_bytes);
+
                // Force-close the channel.
                nodes[0].node.force_close_channel(&OutPoint { txid: tx.txid(), index: 0 }.to_channel_id()).unwrap();
```

Reply:
I believe an `open_channel!()` call will result in a number of calls to notify, and there's a good chance some of them don't necessarily imply a change in the serialized version of the channelmanager (or, maybe more likely, the delay between a notify wakeup and the write getting the write lock may be enough that the next update runs first).